packages/react-juce/src/lib/MacroProperties/util.ts

```diff
@@ -0,0 +1,35 @@
+import { TMacroCall } from "./types";
+
+// don't try too hard, we don't need to support qouted strings containing commas etc.
+export const splitArgs = (a: string) => {
+  const argArray = a.trim().split(/\s*,\s*/);
+  if (argArray.length === 1 && argArray[0] === '') return [];
+  return argArray;
+}
+
+export const getMacroCalls = (
+  s: string,
+  candidates?: string[]
+): TMacroCall[] => {
+  const macroMatcher = String.raw`(${candidates ? `${candidates.join("|")}` : "[a-zA-Z0-9]+"
```

Comment:
Feels a bit like we're writing a code path that we don't even want to use here.

If we pass in the candidates array here by calling `Object.keys(transformFunctionMap)` then the `candidates` argument doesn't need to be optional, or treated as optional (hence no need for the ternary operator here in the regex construction). I think we could invoke an `invariant` here that the candidates array is in fact present.

But, one better, I think we could drop the candidates array altogether and just use the `[a-zA-Z0-9]+` matcher that you have here already. The result will be that we'll return a `macroCalls` array below that _might_ contain function names that we don't actually support. But that's easy enough: in the `reduce` call at the end of `Transform.ts` we can just do

```
const transformMatrix = calls.reduce((acc, { macro: f, args }) => {
  if (!transformFunctionMap.hasOwnProperty(f)) {
    return acc;
  }
  const transform = transformFunctionMap[f](...args);
  return matrix(transform.prod(acc));
}, identity);
```
That feels a little simpler to me. What do you think?

Reply:
Ah yeah, I knew there was a reason I felt a little exposed writing the call to `transformFunctionMap[f]` unchecked, in the case where we haven't passed candidates from the right `Object.keys`, we might not get a function.  I think dropping the candidates arg altogether and checking as you've written is the way to go.